accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,482 @@
+
+In a fully migrated NNBD mode program, an extension with a non-nullable `on` type does not apply to a receiver with a nullable type, and a nullable `on` type means that the `this` value may be `null` inside the extension methods.
+
+During NNBD migration, where non-nullable type may contain `null`, it stays a run-time error if an extension method is called on `null` and a migrated extension's `on` type does not allow null, or an unmigrated extension does not have a trailing `?` on the `on` type. 
+
+During NNBD migration, we will have unsound nullable types like `int*` which should be matched by `extension Wot on int {…}` (in migrated code), and in that case we will need a run-time check to avoid passing `null` to that extension. Unmigrated extensions will still need the trailing `?` to allow getting called with `null` as receiver, but they will apply to nullable types everywhere.
+
+### Semantics of Extension Members
+
+When executing an extension instance member, we stated earlier that the member is invoked with the original receiver as `this` object. We still have to describe how that works, and what the lexical scope is for those members.
+
+Inside an extension method body, `this` is bound to the original receiver, and the static type of `this` is the `on` type (which may contain type variables as usual).
+
+Invocations on `this` use the same extension method resolution as any other code. Most likely  the current extension will be the only one in scope which applies.
+
+Like for a class or mixin member declaration, the names of the extension members, both static and instance, are in the *lexical* scope of the extension member body. That is why `MySmart` above can invoke the static `smartHelper` without prefixing it by the extension name. In the same way, *instance* members are in the lexical scope. 
+
+If an unqualified identifier may lexically resolves to an extension method, the invocation becomes an explicit invocation of that extension method on `this` (which we already know has a compatible type for the extension).
+
+Example:
+
+```dart
+extension MyUnaryNumber on List<Object> {
+  bool get isEven => length.isEven;
+  bool get isOdd => !isEven;
+  static bool isListEven(List<Object> list) => list.isEven;
+}
+```
+
+Here the `list.isEven` will find that `isEven` of `MyUnaryNumber` applies, and unless there are any other extensions in scope, it will call that.
+
+The unqualified `length` of `isEven` is not defined in the current lexical scope, so is equivalent to  `this.length`, which is valid since `List<Object>` has a `length` getter.
+
+The unqualified `isEven` of `isOdd` resolves lexically to the `isEvent` getter above it, so it is equivalent to `MyUnaryNumber(this).isEven`,  even if there are other extensions in scope which define an `isEven` on `List<Object>`.
```

Comment:
I don't think we need anything new, even #328. we can consistently say that a lexically bound identifier refers to the thing in the lexical scope. Whether we do that by a rewrite to `C.foo`, `this.foo` or `Extension(this).foo` is a matter of *what* is in the lexical scope (a static member, an instance member or an extension member).

Reply:
Static extensions are not (except for some variants discussed briefly at the end) related in any way that resembles the subtype and subclass relation of classes, but I think we need a foundation for how to deal with custom bindings of `this` that will also work for classes (and everything else) in a consistent manner.

In a class, I believe that it would be confusing (and inconvenient) if you can call methods of the enclosing class when they are in the lexical scope, but not if they are inherited from a superclass or just added to the interface of the class by the `implements` clause. #328 specifies that we select the instance member of the enclosing class in the case where the interface of that class has a member of the requested name, and that's not something that we would get from a rule that is based solely on lexical scoping.